[GrassHeadd] iP

GrassHeadd opened the pull request:
# Ozymandias User Guide
>My name is Ozymandias, King of Kings;  
>Look on my Works, ye Mighty, and despair!  
>Nothing beside remains. Round the decay   
>
> by [Percy Bysshe Shelley](https://www.poetryfoundation.org/poems/46565/ozymandias) 


All you gotta do is just, ~meet me at the~

1. download it from [here](https://github.com/GrassHeadd/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
- 
If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

GrassHeadd committed:
add Ozymandias.java

GrassHeadd committed:
improve skeletal version of Ozymandias to echo the commands by the user and exits when the user types the command bye

GrassHeadd committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

GrassHeadd committed:
Add the ability to mark tasks as done and change the status back to not done

GrassHeadd committed:
Add support for tracking three types of tasks:
- ToDos: tasks without any date/time attached to it e.g., visit new theme park
- Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
- Events: tasks that start at a specific date/time and ends at a specific date/time
  e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

GrassHeadd committed:
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

GrassHeadd committed:
add simple test case

GrassHeadd committed:
Add support for deleting tasks from the list.

GrassHeadd committed:
fix bug w.r.t runtime error when creating a task that does not have a type

GrassHeadd committed:
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up

GrassHeadd committed:
each the chatbot how to understand dates and times

GrassHeadd committed:
Refactor the code to extract out closely related code as classes.

Extract the following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list

GrassHeadd committed:
Organize the classes into suitable java packages

GrassHeadd committed:
edited readme

GrassHeadd committed:
Merge branch 'add-gradle-support' of https://github.com/GrassHeadd/ip

GrassHeadd committed:
Merge tag 'A-Packages'

* tag 'A-Packages':
  edited readme
  Organize the classes into suitable java packages
  Refactor the code to extract out closely related code as classes.
  each the chatbot how to understand dates and times
  Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up
  fix bug w.r.t runtime error when creating a task that does not have a type

GrassHeadd committed:
Add JUnit tests to test the behavior of the code

GrassHeadd committed:
Package the app as an executable JAR file so that it can be distributed easily.

GrassHeadd committed:
Give users a way to find a task by searching for a keyword in the task description.

GrassHeadd committed:
fix bug for list numbering

GrassHeadd committed:
Tweak the code to comply with a coding standard

GrassHeadd committed:
Add JavaDoc comments

GrassHeadd committed:
Merge pull request #1 from GrassHeadd/Level-9

Level 9

GrassHeadd committed:
Merge pull request #2 from GrassHeadd/A-JavaDoc

Add JavaDoc comments

GrassHeadd committed:
Merge branch 'master' into A-CodingStandard

GrassHeadd committed:
Merge pull request #3 from GrassHeadd/A-CodingStandard

Tweak the code to comply with a coding standard

mihirniyogi commented on src/main/java/Ozymandias/ui/Parser.java:
```diff
@@ -0,0 +1,132 @@
+
+import java.time.format.DateTimeParseException;
+
+public class Parser {
```
btw i think the classes also need javadoc header LOL

GrassHeadd replied:
will add it later!

mihirniyogi commented on src/main/java/Ozymandias/ui/Parser.java:
```diff
@@ -0,0 +1,132 @@
+        }
+
+        if (input.equalsIgnoreCase("bye")) {
+            oz.getUi().greetGoodbye();
+            oz.setExit(true);
+        } else if (input.equalsIgnoreCase("list")) {
+            oz.printTasks();
+        } else if (input.startsWith("mark") || input.startsWith("unmark")) {
+            try {
+                String[] tokens = input.split(" ");
+                int taskId = Integer.parseInt(tokens[1]);
+                if (!oz.getTasks().hasTask(taskId)) {
+                    System.out.println("    THERE IS NO TASK WITH ID " + taskId + "\n");
+                    return;
+                }
+                oz.markTask(taskId, input.startsWith("mark"));
+            } catch (NumberFormatException e) {
+                System.out.println("    TASK ID SHOULD BE AN INTEGER YOU MORON\n.");
+            }
+        } else if (input.startsWith("delete")) {
+            try {
+                String[] tokens = input.split(" ");
+                int taskId = Integer.parseInt(tokens[1]);
+                oz.deleteTask(taskId);
+            } catch (NumberFormatException e) {
+                System.out.println("    TASK ID SHOULD BE AN INTEGER YOU MORON\n");
+            }
+        } else {
+            try {
+                Task newTask = createDifferentTask(input);
+                if (newTask == null) {
+                    System.out.println("    Invalid task or missing details you idiot\n");
+                    return;
+                }
+                oz.addTask(newTask);
+            } catch (IndexOutOfBoundsException e) {
+                System.out.println("    Missing task details or invalid format you banana pants\n");
+            }
+        }
+    }
```
hmm the deep nesting might prove a bit hard to read, u might consider either:
- creating a `Command` class and abstracting out some stuff ... OR
- make it a single try-catch at the start, and catching the errors at the bottom one by one 

this is what i did for my Parser class, just some idea
```java
        if (userInput.startsWith("find")) {
            String[] parts = userInput.split(" ");
            if (parts.length == 1) {
                throw new WrongCommandException("Uh oh! Bob says...the search term cannot be empty.");
            }
            if (parts.length > 2) {
                throw new WrongCommandException("Uh oh! Bob says...the search term must only be 1 word.");
            }
            String searchTerm = parts[1];
            return new FindCommand(searchTerm);
        }
```

EDIT: actually the return null thing right below this is v nice

GrassHeadd replied:
tysm! will refactor out to a command class

mihirniyogi commented on src/main/java/Ozymandias/ui/TaskList.java:
```diff
@@ -0,0 +1,47 @@
+
+    public Task getTask(int id) {return tasks.get(id);}
+
+    public int size() {return tasks.size();}
```
consider changing `size` to `getSize()` to follow the verb-y nature of method naming...
then again there are some built-in java methods that dont follow this rule so up to u

mihirniyogi commented on src/main/java/Ozymandias/ui/TaskList.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public Task removeTask(int id) {
+        if (!tasks.containsKey(id)) {return null;}
```
i like that u followed the 'happy path' rule i.e. handle the error cases at the beginning. kinda wish u did this at other places in the code (e.g. `Parser` class)

mihirniyogi commented on src/main/java/Ozymandias/ui/Task.java:
```diff
@@ -0,0 +1,43 @@
+        this.id = nextId++;
+    }
+
+    public void toggleIsDone() {
+        isDone = !isDone;
+        if (isDone) {
+            System.out.println("    Nice! I've marked this task as done:");
+            System.out.println("      [X] " + taskDetails);
+            System.out.println();
+        } else {
+            System.out.println("    OK, I've marked this task as not done yet:");
+            System.out.println("      [ ] " + taskDetails);
+            System.out.println();
+        }
+    }
```
from an OOP perspective the printing might be better handled by the Ui class u made, same goes for all print statements throughout the program, so that u follow the SRP (single responsibility principle)

mihirniyogi commented on src/main/java/Ozymandias/ui/Events.java:
```diff
@@ -0,0 +1,27 @@
+
+    private static final DateTimeFormatter INPUT_FORMAT  = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    private static final DateTimeFormatter OUTPUT_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy");
+
```
nice naming of constants

mihirniyogi commented on src/main/java/Ozymandias/ui/Events.java:
```diff
@@ -0,0 +1,27 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Events extends Task {
```
personally i would change the plural class naming to singular (e.g. Deadline, Event, Todo) so that it's clear ur not referring to a collection of objects

mihirniyogi commented on src/main/java/Ozymandias/ui/Ozymandias.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    public static void main(String[] args) {
+        Ozymandias oz = new Ozymandias("./data/Ozymandias.txt");
```
consider putting the filepath as a static final String within the `Storage` class, and then tell the Storage class to create file if not exists etc etc

tip is you can use a static block
```java
static {
    // will run even before instantiation
    // e.g. createFileIfNotExists()
}
```

mihirniyogi commented on src/main/java/Ozymandias/ui/Parser.java:
```diff
@@ -0,0 +1,132 @@
+            } catch (DateTimeParseException e) {
+                System.out.println("    Wrong date format for event bread professor!\n");
+                return null;
+            }
```
i feel this is quite clean and organised

mihirniyogi commented on src/main/java/Ozymandias/ui/Ozymandias.java:
```diff
@@ -0,0 +1,137 @@
+     * Runs the Ozymandias application, continuously accept user input
+     * til user exits the program.
+     */
+    public void run() {
+        ui.greetHello();
+        Scanner scanner = new Scanner(System.in);
+
+        while (!isExit) {
+            String input = scanner.nextLine();
+            Parser.handleCommand(input, this);
+            storage.save(tasks);
+        }
+    }
```
clean 💯

mihirniyogi reviewed (commented):
overall quite abstracted out and organised - perhaps few tweaks such as even more abstraction or SRP

GrassHeadd committed:
Update README.md

GrassHeadd committed:
mild refactor of parser

GrassHeadd committed:
more minior refactoring

GrassHeadd committed:
initialise set up for javafx
//todo: refactor code to link with the app

GrassHeadd committed:
edit duke to become ozymandias for javafx

GrassHeadd committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI. Refactor code slightly.

GrassHeadd committed:
Merge pull request #4 from GrassHeadd/Level-10

Merch branch: Level 10

GrassHeadd committed:
parser class: refactor code

GrassHeadd committed:
refactor to rename task directory to tasks

GrassHeadd committed:
did more refactoring

GrassHeadd committed:
minor change to javadoc

GrassHeadd committed:
add assertions

add assertion to ozymandias class at findTask to make sure before the output is return, there should be at least one task if it doesn't go to the edge case of no task

add assertion to ozymandias class at printTasks to make sure the task found are not empty tasks

add assertion to ozymandias class at deleteTasks to make sure the task being delete have been delete

GrassHeadd committed:
Merge pull request #6 from GrassHeadd/A-Assertions

add assertions

GrassHeadd committed:
more refactoring

GrassHeadd committed:
Merge branch 'master' into A-CodeQuality

GrassHeadd committed:
Merge pull request #7 from GrassHeadd/A-CodeQuality

A code quality

GrassHeadd committed:
Merge branch 'master' of https://github.com/GrassHeadd/ip into A-CodeQuality

# Conflicts:
#	src/main/java/Ozymandias/ui/Parser.java

GrassHeadd committed:
Merge branch 'master' of https://github.com/GrassHeadd/ip into A-CodeQuality

# Conflicts:
#	src/main/java/Ozymandias/ui/Parser.java

GrassHeadd committed:
edit weird changes

GrassHeadd committed:
edit the image for chat

GrassHeadd committed:
Merge pull request #8 from GrassHeadd/A-CodeQuality

GrassHeadd committed:
Merge branch 'master' of https://github.com/GrassHeadd/ip

GrassHeadd committed:
slight refactor

GrassHeadd committed:
Provide a way to get reminders about tasks e.g., remind the user about upcoming deadlines

GrassHeadd committed:
Provide a way to get reminders about tasks e.g., remind the user about upcoming deadlines.

GrassHeadd committed:
Merge branch 'master' of https://github.com/GrassHeadd/ip

GrassHeadd committed:
Revert "Provide a way to get reminders about tasks e.g., remind the user about upcoming deadlines."

This reverts commit 8c7dd7a81c6870598c7baaa68aa5e0f9f483b7a3.

GrassHeadd committed:
refactor slightly

GrassHeadd committed:
more refactor

GrassHeadd committed:
edit toggle

GrassHeadd committed:
more refactoring

GrassHeadd committed:
minor refactor

GrassHeadd committed:
Merge pull request #9 from GrassHeadd/B-Reminder

minor refactor

GrassHeadd committed:
debug attempt

GrassHeadd committed:
fix issue

GrassHeadd committed:
fix attempt

GrassHeadd committed:
remove entry point in ozymandias.java

GrassHeadd committed:
edit behaviour of remind

GrassHeadd committed:
remove debugging statements

GrassHeadd committed:
add ui

GrassHeadd committed:
Update README.md

GrassHeadd committed:
minor refinements

GrassHeadd committed:
Update README.md

GrassHeadd committed:
add javadocs and refactor

GrassHeadd committed:
Update README.md

GrassHeadd committed:
minor edit

GrassHeadd committed:
fix issue with storage load

GrassHeadd committed:
fix bug
